sql/src/SqlCompile.sk

```diff
@@ -567,15 +569,31 @@ mutable class Compiler{
           )
         }
 
+      | P.SelectExpr(P.Check(_), _) -> void
       | p @ P.SelectExpr _ -> expandedSParams.push(p)
       }
     };
 
     expandedSParams.toArray().map(param -> {
-      this.compileExpr(context, param.expr)
+      this.compileExpr(context, param.expr, param.alias)
     })
   }
 
+  mutable fun compileSideEffectParams(
+    context: mutable SKStore.Context,
+    selectParams: Array<P.SelectResult>,
+  ): Array<CGExpr> {
+    compiledExprs = mutable Vector[];
+    for (_ => param in selectParams) {
+      param match {
+      | P.SelectExpr(check @ P.Check(_), alias) ->
```

Comment:
I'll check this (no pun intended). I'm pretty sure `1 + CHECK(0)` will fail to evaluate. This is why I introduced ASuccess() and have the throw clauses everywhere.

Reply:
I checked it on your branch while reviewing, and get the following output:
```
skdb> select 1 + check(0);
select 1 + check(0);
^
|
 ----- ERROR
Error: line 1, character 0:
Failed constraint: 0
```

Reply:
That's disappointing. `select 1 + check(1)` doesn't fail either - I was expecting that to produce ASuccess() which BinOp would hopefully choke on. I'll dig in and see what's happening.

Reply:
Fixed. This was really dumb. I introduced ASuccess for exactly this reason: to deal with CHECK expressions as subexpressions. But I didn't fail on the ASuccess() branch. 🤦‍♂️

Reply:
Out of curiosity: why make this a runtime error instead of a static compile-time check?  Since it's a purely syntactic property, we ought to be able to just inspect the AST upfront instead of baking it into the semantics like this.

The current setup produces some (to me) unintuitive behaviour discrepancy between passing and failing checks in subexpressions.

i.e.
```
skdb> select 1 + check(1);
select 1 + check(1);
^
|
 ----- ERROR
Error: line 1, character 0:
SELECT error: parameter expression contains a side effect
skdb> select 1 + check(0);
select 1 + check(0);
^
|
 ----- ERROR
Error: line 1, character 0:
Failed constraint: 0
```

(sorry not to elaborate on/express this concern more clearly in my initial comments... feel free to ignore if it's a pain to change the dynamic checks out for static ones)

Reply:
> Out of curiosity: why make this a runtime error instead of a static compile-time check?

Good question. To me it's all runtime as the compilation happens at runtime. So I didn't give it much thought. My intuition said handle at eval as that's a deeper more fundamental layer and so will stand up to code change better, rather than putting in a frontend check and assuming afterward all is good and validated.

But...

> The current setup produces some (to me) unintuitive behaviour discrepancy between passing and failing checks in subexpressions.

This is legit.

I'll take a look at introducing compile time checking.

Reply:
Yep, np, here we go:

```
# ./build/skdb <<< "select 1 + check(1);"
select 1 + check(1);
^
|
 ----- ERROR
Error: line 1, character 0:
CHECK expressions cannot be nested
1 # ./build/skdb <<< "select 1 + check(0);"
select 1 + check(0);
^
|
 ----- ERROR
Error: line 1, character 0:
CHECK expressions cannot be nested
1 # ./build/skdb <<< "select check(0);"
select check(0);
^
|
 ----- ERROR
Error: line 1, character 0:
Failed constraint: 0
1 # ./build/skdb <<< "select check(1);"
# 
```

I'll push the commit in a sec. Best of both worlds, we have the compile time check and this is also handled at eval time just in case.